File: packages/nerva-py/nerva/router/llm.py
"""LLM router — classify intent by asking an LLM to select a handler (N-610).

Sends the full handler catalog (names + descriptions) to an LLM and parses
the JSON response to produce an IntentResult. Falls back gracefully when the
LLM returns unparseable output.
"""

from __future__ import annotations

import json
import logging
import re
from dataclasses import dataclass
from typing import TYPE_CHECKING, Protocol

from nerva.router import HandlerCandidate, IntentResult

if TYPE_CHECKING:
    from nerva.context import ExecContext

__all__ = [
    "LLMFunc",
    "LLMRouter",
    "LLMRouterConfig",
]

_log = logging.getLogger(__name__)

# -- Constants ---------------------------------------------------------------

NO_MATCH_INTENT: str = "unknown"
NO_MATCH_CONFIDENCE: float = 0.0
LLM_INTENT: str = "llm"
DEFAULT_CONFIDENCE: float = 0.5
"""Confidence assigned when the LLM omits a confidence value."""

JSON_EXTRACT_PATTERN: re.Pattern[str] = re.compile(r"\{[^{}]*\}", re.DOTALL)
"""Regex to locate a JSON object in potentially noisy LLM output."""

DEFAULT_SYSTEM_PROMPT: str = (
    "You are an intent classifier. Given the handler catalog and a user message, "
    "select the single best handler.\n"
    "Respond ONLY with JSON: {\"handler\": \"<name>\", \"confidence\": <0.0-1.0>}\n"
    "If no handler fits, respond: {\"handler\": \"\", \"confidence\": 0.0}"
)
"""Default system prompt template sent to the LLM."""


# -- LLM function protocol --------------------------------------------------


class LLMFunc(Protocol):
    """Async function that sends a system+user prompt pair to an LLM.

    Args:
        system_prompt: Instructions for the LLM.
        user_prompt: The user-facing query to classify.

    Returns:
        Raw text response from the LLM.
    """

    async def __call__(self, system_prompt: str, user_prompt: str) -> str: ...


# -- Config ------------------------------------------------------------------


@dataclass(frozen=True)
class LLMRouterConfig:
    """Configuration for the LLM router.

    Attributes:
        system_prompt: System prompt template for the LLM.
        fallback_confidence: Confidence when LLM omits the field.
    """

    system_prompt: str = DEFAULT_SYSTEM_PROMPT
    fallback_confidence: float = DEFAULT_CONFIDENCE


# -- Internal value object ---------------------------------------------------


@dataclass(frozen=True)
class _RegisteredHandler:
    """A handler registered with the LLM router.

    Attributes:
        name: Handler identifier.
        description: Human-readable description shown to the LLM.
    """

    name: str
    description: str


# -- Router ------------------------------------------------------------------


class LLMRouter:
    """Route messages by asking an LLM to pick the best handler from a catalog.

    Builds a prompt containing all registered handler names and descriptions,
    sends it to the injected LLM function, and parses the JSON response into
    an IntentResult. Falls back to an empty result when the LLM returns
    invalid output.

    Args:
        llm: Async LLM function that accepts (system_prompt, user_prompt).
        config: Optional router configuration overrides.
    """

    def __init__(
        self,
        llm: LLMFunc,
        config: LLMRouterConfig | None = None,
    ) -> None:
        self._llm = llm
        self._config = config or LLMRouterConfig()
        self._handlers: list[_RegisteredHandler] = []

    # -- Public API ----------------------------------------------------------

    async def register(self, name: str, description: str) -> None:
        """Register a handler for LLM-based routing.

        Args:
            name: Unique handler identifier.
            description: Human-readable description shown to the LLM.

        Raises:
            ValueError: If *name* is empty or already registered.
            ValueError: If *description* is empty or whitespace-only.
        """
        if not name:
            raise ValueError("Handler name must not be empty")
        if not description or not description.strip():
            raise ValueError("Handler description must not be empty")
        if _find_handler(self._handlers, name) is not None:
            raise ValueError(f"Handler '{name}' is already registered")

        self._handlers.append(_RegisteredHandler(name=name, description=description))

    async def classify(self, message: str, ctx: ExecContext) -> IntentResult:
        """Classify a message by asking the LLM to select a handler.

        Args:
            message: Raw user message text.
            ctx: Execution context (forwarded for observability).

        Returns:
            IntentResult with the LLM's handler selection and confidence.
        """
        if not message or not message.strip():
            return _empty_result()
        if not self._handlers:
            return _empty_result()

        user_prompt = _build_user_prompt(self._handlers, message)
        raw_response = await self._llm(self._config.system_prompt, user_prompt)

        return self._parse_response(raw_response)

    # -- Private -------------------------------------------------------------

    def _parse_response(self, raw: str) -> IntentResult:
        """Parse the LLM's JSON response into an IntentResult.

        Tries full-string JSON parse first, then regex extraction.
        Falls back to an empty result on any parse failure.

        Args:
            raw: Raw text response from the LLM.

        Returns:
            Parsed IntentResult, or empty result on failure.
        """
        parsed = _try_parse_json(raw)
        if parsed is None:
            parsed = _regex_extract_json(raw)
        if parsed is None:
            _log.warning("LLM returned unparseable response: %s", raw[:200])
            return _empty_result()

        return self._build_result_from_parsed(parsed)

    def _build_result_from_parsed(self, parsed: dict[str, object]) -> IntentResult:
        """Convert parsed JSON into an IntentResult.

        Validates the handler name exists in the catalog and clamps confidence.

        Args:
            parsed: Parsed JSON dict from the LLM response.

        Returns:
            IntentResult with the selected handler, or empty if invalid.
        """
        handler_name = str(parsed.get("handler", "")).strip()
        if not handler_name:
            return _empty_result()

        if _find_handler(self._handlers, handler_name) is None:
            _log.warning("LLM selected unknown handler: %s", handler_name)
            return _empty_result()

        confidence = _extract_confidence(parsed, self._config.fallback_confidence)
        candidate = HandlerCandidate(
            name=handler_name,
            score=confidence,
            reason="selected by LLM",
        )
        return IntentResult(
            intent=LLM_INTENT,
            confidence=confidence,
            handlers=[candidate],
            raw_scores={handler_name: confidence},
        )


# -- Pure helpers ------------------------------------------------------------


def _build_user_prompt(handlers: list[_RegisteredHandler], message: str) -> str:
    """Build the user prompt containing the handler catalog and message.

    Args:
        handlers: All registered handlers.
        message: The user message to classify.

    Returns:
        Formatted prompt string for the LLM.
    """
    catalog_lines = [f"- {h.name}: {h.description}" for h in handlers]
    catalog_text = "\n".join(catalog_lines)
    return f"Available handlers:\n{catalog_text}\n\nUser message: {message}"


def _find_handler(
    handlers: list[_RegisteredHandler], name: str
) -> _RegisteredHandler | None:
    """Look up a registered handler by name.

    Args:
        handlers: List of registered handlers.
        name: Handler name to search for.

    Returns:
        The matching handler, or None if not found.
    """
    for handler in handlers:
        if handler.name == name:
            return handler
    return None


def _try_parse_json(text: str) -> dict[str, object] | None:
    """Attempt to parse the entire text as a JSON object.

    Args:
        text: Candidate JSON string.

    Returns:
        Parsed dict if successful, None otherwise.
    """
    stripped = text.strip()
    if not stripped:
        return None
    try:
        result = json.loads(stripped)
        if isinstance(result, dict):
            return result
    except (json.JSONDecodeError, ValueError):
        pass
    return None


def _regex_extract_json(text: str) -> dict[str, object] | None:
    """Use regex to find and parse the first JSON object in text.

    Args:
        text: Noisy LLM output that may contain a JSON object.

    Returns:
        Parsed dict from the first valid JSON match, or None.
    """
    for match in JSON_EXTRACT_PATTERN.finditer(text):
        parsed = _try_parse_json(match.group())
        if parsed is not None:
            return parsed
    return None


def _extract_confidence(parsed: dict[str, object], fallback: float) -> float:
    """Extract and clamp the confidence value from parsed JSON.

    Args:
        parsed: Parsed LLM response dict.
        fallback: Value to use when confidence is missing or invalid.

    Returns:
        Confidence clamped to [0.0, 1.0].
    """
    raw = parsed.get("confidence")
    if raw is None:
        return fallback
    try:
        value = float(raw)
    except (TypeError, ValueError):
        return fallback
    return max(0.0, min(1.0, value))


def _empty_result() -> IntentResult:
    """Build an empty IntentResult when nothing matched.

    Returns:
        IntentResult with zero confidence and no handlers.
    """
    return IntentResult(
        intent=NO_MATCH_INTENT,
        confidence=NO_MATCH_CONFIDENCE,
        handlers=[],
    )

File: packages/nerva-js/src/router/llm.ts
/**
 * LLM router — classify intent by asking an LLM to select a handler.
 *
 * Builds a prompt containing the full handler catalog, sends it to the
 * provided LLM function, and parses the JSON response into an
 * {@link IntentResult}. Falls back gracefully on invalid JSON.
 *
 * @module router/llm
 */

import type { ExecContext } from "../context.js";
import type { IntentResult, IntentRouter } from "./index.js";
import { createHandlerCandidate, createIntentResult } from "./index.js";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/**
 * Async function that sends a system prompt and user prompt to an LLM
 * and returns the raw completion text.
 */
export type LLMFn = (systemPrompt: string, userPrompt: string) => Promise<string>;

/**
 * Shape of the JSON the LLM is expected to return.
 */
interface LLMRouterResponse {
  readonly handler: string;
  readonly confidence: number;
}

// ---------------------------------------------------------------------------
// Constants
// ---------------------------------------------------------------------------

/** Intent label for LLM-routed matches. */
const LLM_INTENT = "llm";

/** Intent label when nothing matched. */
const NO_MATCH_INTENT = "unknown";

/** Confidence for empty or failed results. */
const NO_MATCH_CONFIDENCE = 0.0;

/** Default confidence assigned when the LLM omits one. */
const DEFAULT_LLM_CONFIDENCE = 0.5;

/** Regex to locate a JSON object in noisy LLM output. */
const JSON_EXTRACT_PATTERN = /\{[^{}]*\}/s;

// ---------------------------------------------------------------------------
// Registered handler
// ---------------------------------------------------------------------------

/** A handler registered in the catalog. */
interface RegisteredHandler {
  readonly name: string;
  readonly description: string;
}

// ---------------------------------------------------------------------------
// LLMRouter
// ---------------------------------------------------------------------------

/**
 * Routes messages by asking an LLM to select the best handler from a catalog.
 *
 * Builds a system prompt listing all registered handlers and their
 * descriptions, then sends the user message as a user prompt. The LLM
 * must return JSON with `handler` and `confidence` fields.
 *
 * On parse failure or unknown handler, returns an "unknown" intent.
 *
 * @example
 * ```ts
 * const router = new LLMRouter(myLlmFn);
 * router.register("flights", "Book airline flights");
 * router.register("hotels", "Reserve hotel rooms");
 * const result = await router.classify("I need a flight to Paris", ctx);
 * ```
 */
export class LLMRouter implements IntentRouter {
  private readonly _llm: LLMFn;
  private readonly _handlers: RegisteredHandler[] = [];

  /**
   * @param llm - Async function that calls an LLM with system and user prompts.
   */
  constructor(llm: LLMFn) {
    this._llm = llm;
  }

  /**
   * Register a handler in the catalog.
   *
   * @param name - Unique handler identifier.
   * @param description - Human-readable description shown to the LLM.
   * @throws {Error} If name is empty or already registered.
   * @throws {Error} If description is empty or whitespace-only.
   */
  register(name: string, description: string): void {
    if (!name) {
      throw new Error("Handler name must not be empty");
    }
    if (!description || !description.trim()) {
      throw new Error("Handler description must not be empty");
    }
    if (this._handlers.some((h) => h.name === name)) {
      throw new Error(`Handler '${name}' is already registered`);
    }
    this._handlers.push({ name, description });
  }

  /**
   * Classify intent by asking the LLM to select a handler.
   *
   * @param message - Raw user message text.
   * @param _ctx - Execution context (forwarded for observability; not used directly).
   * @returns IntentResult with the LLM's selected handler and confidence.
   */
  async classify(message: string, _ctx: ExecContext): Promise<IntentResult> {
    if (!message || !message.trim()) {
      return emptyResult();
    }
    if (this._handlers.length === 0) {
      return emptyResult();
    }

    const systemPrompt = buildSystemPrompt(this._handlers);
    const rawResponse = await this._llm(systemPrompt, message);
    return parseResponse(rawResponse, this._handlers);
  }
}

// ---------------------------------------------------------------------------
// Pure helpers
// ---------------------------------------------------------------------------

/**
 * Build the system prompt listing all available handlers.
 *
 * @param handlers - Registered handler catalog.
 * @returns A system prompt instructing the LLM to return JSON.
 */
function buildSystemPrompt(handlers: readonly RegisteredHandler[]): string {
  const catalog = handlers
    .map((h) => `- ${h.name}: ${h.description}`)
    .join("\n");

  return [
    "You are an intent classifier. Given a user message, select the best handler from the catalog below.",
    "Return ONLY a JSON object with two fields: \"handler\" (the handler name) and \"confidence\" (a number between 0.0 and 1.0).",
    "Do not include any other text.",
    "",
    "Available handlers:",
    catalog,
  ].join("\n");
}

/**
 * Parse the LLM response into an IntentResult.
 *
 * Tries full JSON parse first, then regex extraction. Falls back to
 * an empty result on invalid JSON or unknown handler names.
 *
 * @param rawResponse - Raw text from the LLM.
 * @param handlers - Registered handlers for validation.
 * @returns IntentResult with the parsed handler, or unknown on failure.
 */
function parseResponse(
  rawResponse: string,
  handlers: readonly RegisteredHandler[],
): IntentResult {
  const parsed = extractJson(rawResponse);
  if (parsed === null) {
    return emptyResult();
  }

  const handlerName = parsed.handler;
  if (!handlers.some((h) => h.name === handlerName)) {
    return emptyResult();
  }

  const confidence = clampConfidence(parsed.confidence);
  const candidate = createHandlerCandidate(handlerName, confidence, "llm selection");
  return createIntentResult(LLM_INTENT, confidence, [candidate], {
    [handlerName]: confidence,
  });
}

/**
 * Extract and validate a JSON object from potentially noisy LLM output.
 *
 * @param raw - Raw LLM response text.
 * @returns Parsed response with handler and confidence, or null on failure.
 */
function extractJson(raw: string): LLMRouterResponse | null {
  const stripped = raw.trim();
  if (stripped === "") return null;

  const fullParse = tryParseResponse(stripped);
  if (fullParse !== null) return fullParse;

  const match = JSON_EXTRACT_PATTERN.exec(stripped);
  if (match !== null) {
    return tryParseResponse(match[0]);
  }

  return null;
}

/**
 * Try to parse text as a valid LLM router response.
 *
 * @param text - Candidate JSON string.
 * @returns Validated response or null if invalid.
 */
function tryParseResponse(text: string): LLMRouterResponse | null {
  try {
    const data: unknown = JSON.parse(text);
    if (typeof data !== "object" || data === null || Array.isArray(data)) {
      return null;
    }
    const record = data as Record<string, unknown>;
    const handler = record["handler"];
    if (typeof handler !== "string" || handler === "") {
      return null;
    }
    const confidence = typeof record["confidence"] === "number"
      ? record["confidence"]
      : DEFAULT_LLM_CONFIDENCE;
    return { handler, confidence };
  } catch {
    return null;
  }
}

/**
 * Clamp a confidence value to the [0.0, 1.0] range.
 *
 * @param value - Raw confidence from the LLM.
 * @returns Clamped confidence between 0.0 and 1.0.
 */
function clampConfidence(value: number): number {
  if (!Number.isFinite(value)) return DEFAULT_LLM_CONFIDENCE;
  return Math.max(0.0, Math.min(1.0, value));
}

/**
 * Build an empty IntentResult when nothing matched.
 *
 * @returns IntentResult with zero confidence and no handlers.
 */
function emptyResult(): IntentResult {
  return createIntentResult(NO_MATCH_INTENT, NO_MATCH_CONFIDENCE, []);
}
